Use bleach instead of python-markdown's Safe Mode

ryanlerch opened the pull request:
Previously, we used the Safe Mode of python-markdown to
replace HTML tags in the markdown input from the user. It
replaced all text with a Raw HTML not allowed message.

However, the Safe Mode feature in python-markdown was deprecated,
and they recomended that Bleach be used instead.

This commit passes the markdown input from the user through
Bleach -- before it renders the markdown to html -- escaping all
HTML tags.

ryanlerch committed:
Use bleach instead of python-markdown's Safe Mode

Previously, we used the Safe Mode of python-markdown to
replace HTML tags in the markdown input from the user. It
replaced all text with a Raw HTML not allowed message.

However, the Safe Mode feature in python-markdown was deprecated,
and they recomended that Bleach be used instead.

This commit passes the markdown input from the user through
Bleach -- before it renders the markdown to html -- escaping all
HTML tags.

ryanlerch commented on bodhi/server/util.py:
```diff
@@ -297,9 +298,23 @@ def hostname(context=None):
+    Return HTML from a markdown string
+
+    Args:
+        context:
```
Did not know how to describe `context` here. Any thoughts?

Do we even need to pass context to this method?

pypingou replied:
I believe passing it is required, but we could s/context/_/ in the function definition

ryanlerch replied:
ah! forgot that this method was used in the templates too, where context is passed automatically!

do you think we should just leave it as context, since all the others use it?

pypingou replied:
+1 to leave it

bowlofeggs replied:
```context``` is a ```mako.runtime.Context```. I just stuck a log statement here to find out the type, like this:

```
log.info('type(context): %s' % type(context))
```

pypingou commented on bodhi/server/util.py:
```diff
@@ -297,9 +298,23 @@ def hostname(context=None):
+    # previously, we used the Safe Mode in python-markdown to strip all HTML
+    # tags. Safe Mode is deprecated, so we now use Bleach to sanitize all HTML
+    # tags before running it through the markdown parser
+    sanitized_text = bleach.clean(text, tags=[])
```
Note that bleach broke it's API in 2.0, so in pagure we've had to do some tricks to work pre and pose 2.0: https://pagure.io/pagure/pull-request/2269

ryanlerch replied:
I wonder what version of bleach is default in prod where we run bodhi... Hopefully we can avoid this dance...

ralphbean replied:
Looks like python-bleach-1.4.1-2.el7 is in epel7, which is what bodhi runs on.  (Confirmed, that's exactly the version available to the `bodhi03` VM).

(Someone could always put the latest bleach > 2.0 in the epel7-infra tag).

ryanlerch replied:
python-bleach is still version 1.x in Fedora too:

https://apps.fedoraproject.org/packages/python-bleach

are we able to just add `bleach<2` in the requirements.txt to solve this?

bowlofeggs replied:
Since bodhi itself is in EPEL, we shouldn't require a newer bleach than is available in EPEL. However, Bodhi is also in Rawhide. Rawhide currently has bleach 1.5, but I would assume it could get updated at any time.

ryanlerch replied:
Just FYI, we didnt avoid this dance for long... Rawhide bleach is now updated to 2. see #1718

pypingou reviewed (commented):
Couple of comments but overall 👍 for me

bowlofeggs commented on bodhi/tests/server/functional/test_generic.py:
```diff
@@ -195,10 +195,11 @@ def test_markdown(self):
 
     def test_markdown_with_html(self):
         res = self.app.get('/markdown', {'text': '<b>bold</b>'}, status=200)
+        print res
```
Let's drop this print statement.

bowlofeggs reviewed (changes requested):
A couple more things:

* Let's put the type on the context (see my earlier comment).
* Let's put a note in the release_notes.rst file that bleach is a new dependency.

bowlofeggs commented:
On Tue, 2017-05-30 at 20:02 -0700, Ryan Lerch wrote:
> are we able to just add bleach<2 in the requirements.txt to solve
> this?

Sure, that works for me. We can burn the bleach == 2 bridge when we get
to it.

ryanlerch commented:
* added a note in the release notes
* updated requirements.txt to require bleach<2
* added the type to the context in the docblock
* removed one of my now infamous print statements i like to leave lying around.

pypingou commented:
Looks fine to me

bowlofeggs reviewed (approved):
@ryanlerch you are a hero
